src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -964,22 +964,21 @@ public Rope withEncodingSameEncoding(Rope rope, Encoding encoding, CodeRange cod
 
         @Specialization(guards = {
                 "rope.getEncoding() != encoding",
-                "rope.getCodeRange() == codeRange",
-                "rope.canFastEncode(encoding, codeRange)"
+                "!rope.canFastEncode(encoding, codeRange)"
         })
-        public Rope withEncodingSameCodeRange(Rope rope, Encoding encoding, CodeRange codeRange) {
-            return rope.withEncoding(encoding, codeRange);
+        public Rope withSlowEncoding(Rope rope, Encoding encoding, CodeRange codeRange,
+                @Cached("create()") MakeLeafRopeNode makeLeafRopeNode,
+                @Cached("create()") RopeNodes.BytesNode bytesNode) {
+            return makeLeafRopeNode.executeMake(bytesNode.execute(rope), encoding, codeRange, NotProvided.INSTANCE);
         }
 
         @Specialization(guards = {
                 "rope.getEncoding() != encoding",
                 "rope.getCodeRange() == codeRange",
-                "!rope.canFastEncode(encoding, codeRange)"
+                "rope.canFastEncode(encoding, codeRange)"
```

Comment:
I'm confused by this `canFastEncode` method. As far as I can tell, it mostly does the code range check that existed before that method was added. In this case, you still have the code range guard and then the `canFastEncode` guard.

What problem are you addressing with this? I think the notion of fast encoding is a property of the just the encoding and code range, which shouldn't be rope-type dependent.

Reply:
It's a property of the encoding, the code range, and the rope. Whether it should be a property of the rope as well is a good question, but some Rope subclasses will always throw an exception, while others threw an exception depending on the encoding and range.